cmd/aks-periscope/aks-periscope.go

```diff
@@ -32,40 +31,183 @@ func main() {
+			containerLogsCollectorContainerD.GetName(): containerLogsCollectorContainerD,
+			systemLogsCollector.GetName():              systemLogsCollector,
+			networkOutboundCollector.GetName():         networkOutboundCollector,
+			ipTablesCollector.GetName():                ipTablesCollector,
+			nodeLogsCollector.GetName():                nodeLogsCollector,
+			dnsCollector.GetName():                     dnsCollector,
+			kubeObjectsCollector.GetName():             kubeObjectsCollector,
+			kubeletCmdCollector.GetName():              kubeletCmdCollector,
+			systemPerfCollector.GetName():              systemPerfCollector,
+			helmCollector.GetName():                    helmCollector,
+			osmCollector.GetName():                     osmCollector,
+		})
+
+	//diagnosers
+	//NOTE currently the collector instance objects are shared between the collector itself and things which use it as a dependency
+	networkConfigDiagnoser := diagnoser.NewNetworkConfigDiagnoser(dnsCollector, kubeletCmdCollector, selectedExporters)
+	networkOutboundDiagnoser := diagnoser.NewNetworkOutboundDiagnoser(networkOutboundCollector, selectedExporters)
+	selectedDiagnosers := selectDiagnosers(
+		map[string]interfaces.Diagnoser{
+			networkConfigDiagnoser.GetName():   networkConfigDiagnoser,
+			networkOutboundDiagnoser.GetName(): networkOutboundDiagnoser,
+		})
+
+	return selectedCollectors, selectedDiagnosers, selectedExporters
+}
+
+// selectCollectors select the collectors to run
+func selectCollectors(allCollectorsByName map[string]interfaces.Collector) []interfaces.Collector {
 	collectors := []interfaces.Collector{}
-	containerLogsCollector := collector.NewContainerLogsCollector(exporter)
-	networkOutboundCollector := collector.NewNetworkOutboundCollector(5, exporter)
-	dnsCollector := collector.NewDNSCollector(exporter)
-	kubeObjectsCollector := collector.NewKubeObjectsCollector(exporter)
-	systemLogsCollector := collector.NewSystemLogsCollector(exporter)
-	ipTablesCollector := collector.NewIPTablesCollector(exporter)
-	nodeLogsCollector := collector.NewNodeLogsCollector(exporter)
-	kubeletCmdCollector := collector.NewKubeletCmdCollector(exporter)
-	systemPerfCollector := collector.NewSystemPerfCollector(exporter)
-	helmCollector := collector.NewHelmCollector(exporter)
-	osmCollector := collector.NewOsmCollector(exporter)
-
-	collectors = append(collectors, containerLogsCollector)
-	collectors = append(collectors, dnsCollector)
-	collectors = append(collectors, kubeObjectsCollector)
-	collectors = append(collectors, networkOutboundCollector)
 
-	if contains(collectorList, "connectedCluster") {
-		collectors = append(collectors, helmCollector)
+	//read list of collectors that are enabled
+	var enabledCollectorNames []string
+
+	//TODO try get partners to move from COLLECTOR_LIST to use ENABLED_COLLECTORS instead, for now COLLECTOR_LIST takes precedence if defined
+	collectorList, found := os.LookupEnv("COLLECTOR_LIST")
```

Comment:
Nice! Proposal sounds good, second part might be tricky

Reply:
I see 2 basic options for now:
- do what the users was expected to do: check the cluster stack and decide whether to run or not
- always run, but return nothing if there is an error

Reply:
I disagree. For now we should retain the current behaviour - users might not want all collectors / diagnosers / exporters to execute, so even if we can confirm the current cluster stack and test that all c / d / e 's can run and safely err on every environment without side effect, it is useful to retain support for manual override.

COLLECTOR_LIST needs to be retained for backcompat with existing consumers. We can mark it deprecated though.

Reply:
it's just an env variable, we can drop it safely this won't break anything

Reply:
It won't break anything for us but will force changes on partners, which I have been trying to minimize. It is also self contained and easy to remove later

Reply:
I don't see the point with issue here. The worst it can do is gather more files, the backcompat you are talking about is ensured. There is no change needed for the partners.
Let's tackle this in another PR if you want